nucypher/utilities/porter/porter.py

```diff
@@ -0,0 +1,138 @@
+from nucypher.utilities.logging import Logger
+from nucypher.utilities.porter.control.controllers import PorterCLIController
+from nucypher.utilities.porter.control.interfaces import PorterInterface
+
+
+class Porter(Learner):
+
+    BANNER = r"""
+
+ ______
+(_____ \           _
+ _____) )__   ____| |_  ____  ____
+|  ____/ _ \ / ___)  _)/ _  )/ ___)
+| |   | |_| | |   | |_( (/ /| |
+|_|    \___/|_|    \___)____)_|
+
+the conduit between web apps and the nucypher network
+"""
+
+    APP_NAME = "Porter"
+
+    _SHORT_LEARNING_DELAY = 2
+    _LONG_LEARNING_DELAY = 30
+    _ROUNDS_WITHOUT_NODES_AFTER_WHICH_TO_SLOW_DOWN = 25
+
+    DEFAULT_PORTER_HTTP_PORT = 9155
+
+    _interface_class = PorterInterface
+
+    def __init__(self,
+                 domain: str = None,
+                 registry: BaseContractRegistry = None,
+                 controller: bool = True,
+                 *args, **kwargs):
+        self.federated_only = False  # TODO Start with non-federated for now
+        self.registry = registry or InMemoryContractRegistry.from_latest_publication(network=domain)
+
+        super().__init__(save_metadata=True, domain=domain, *args, **kwargs)
+
+        self.log = Logger(self.__class__.__name__)
+
+        # Controller Interface
+        self.interface = self._interface_class(porter=self)
+        self.controller = NO_CONTROL_PROTOCOL
+        if controller:
+            # TODO need to understand this better - only made it analogous to what was done for characters
+            self.make_cli_controller()
+        self.log.info(self.BANNER)
+
+    def make_cli_controller(self, crash_on_error: bool = False):
+        controller = PorterCLIController(app_name=self.APP_NAME,
+                                         crash_on_error=crash_on_error,
+                                         interface=self.interface)
+        self.controller = controller
+        return controller
+
+    def make_rpc_controller(self, crash_on_error: bool = False):
+        controller = JSONRPCController(app_name=self.APP_NAME,
+                                       crash_on_error=crash_on_error,
+                                       interface=self.interface)
+
+        self.controller = controller
+        return controller
+
+    def make_web_controller(self, crash_on_error: bool = False):
+        controller = WebController(app_name=self.APP_NAME,
+                                   crash_on_error=crash_on_error,
+                                   interface=self._interface_class(porter=self))
+        self.controller = controller
+
+        # Register Flask Decorator
+        porter_flask_control = controller.make_control_transport()
+
+        #
+        # Porter Control HTTP Endpoints
+        #
+        @porter_flask_control.route('/get_ursulas', methods=['GET'])
+        def get_ursulas() -> Response:
+            """
+            Porter control endpoint for sampling Ursulas on behalf of Alice.
+            """
```

Comment:
Style: These docstrings can all be one-liners.

Reply:
✔️